src-java/nbworker-topology/nbworker-storm-topology/src/main/java/org/openkilda/wfm/topology/nbworker/validators/PathValidator.java

```diff
@@ -0,0 +1,580 @@
+import java.util.stream.Collectors;
+
+public class PathValidator {
+
+    private static final String LATENCY = "latency";
+    private static final String LATENCY_TIER_2 = "latency tier 2";
+    private final IslRepository islRepository;
+    private final FlowRepository flowRepository;
+    private final SwitchPropertiesRepository switchPropertiesRepository;
+    private final SwitchRepository switchRepository;
+    private Map<IslEndpoints, Optional<Isl>> islCache;
+    private Map<SwitchId, Optional<Switch>> switchCache;
+
+    public PathValidator(IslRepository islRepository,
+                         FlowRepository flowRepository,
+                         SwitchPropertiesRepository switchPropertiesRepository,
+                         SwitchRepository switchRepository) {
+        this.islRepository = islRepository;
+        this.flowRepository = flowRepository;
+        this.switchPropertiesRepository = switchPropertiesRepository;
+        this.switchRepository = switchRepository;
+    }
+
+    /**
+     * Validates whether it is possible to create a path with the given parameters. When there are obstacles, this
+     * validator returns all errors found on each segment. For example, when there is a path 1-2-3-4 and there is
+     * no link between 1 and 2, no sufficient latency between 2 and 3, and not enough bandwidth between 3-4, then
+     * this validator returns all 3 errors.
+     *
+     * @param pathValidationData path parameters to validate.
+     * @return a response object containing the validation result and errors if any
+     */
+    public PathValidationResult validatePath(PathValidationData pathValidationData) {
+        initCache();
+        Set<String> result = pathValidationData.getPathSegments().stream()
+                .map(segment -> executeValidations(
+                        InputData.of(pathValidationData, segment),
+                        getPerSegmentValidations(pathValidationData)))
+                .flatMap(Set::stream)
+                .collect(Collectors.toSet());
+
+        result.addAll(executeValidations(InputData.of(pathValidationData),
+                getPerPathValidations(pathValidationData)));
+
+        return PathValidationResult.builder()
+                .isValid(result.isEmpty())
+                .errors(new LinkedList<>(result))
+                .build();
+    }
+
+    private void initCache() {
+        islCache = new HashMap<>();
+        switchCache = new HashMap<>();
+    }
+
+    private Optional<Isl> findIslByEndpoints(SwitchId srcSwitchId, int srcPort, SwitchId destSwitchId, int destPort) {
+        IslEndpoints endpoints = new IslEndpoints(srcSwitchId.toString(), srcPort, destSwitchId.toString(), destPort);
+        if (!islCache.containsKey(endpoints)) {
+            islCache.put(endpoints,
+                    islRepository.findByEndpoints(srcSwitchId, srcPort, destSwitchId, destPort));
+        }
+        return islCache.get(endpoints);
+    }
```

Comment:
hint: You can use lazy map 
https://github.com/telstra/open-kilda/blob/36a896231e8fb9c4c70c66a6ef20e54f4d4e8a1d/src-java/base-topology/base-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowPathBuilder.java#L214-L220

Reply:
done